rust/wal3/src/gc.rs

```diff
@@ -0,0 +1,313 @@
+use std::ops::Add;
+
+use chroma_storage::Storage;
+use setsum::Setsum;
+
+use crate::{
+    deserialize_setsum, serialize_setsum, Error, Fragment, LogPosition, Manifest, ScrubError,
+    Snapshot, SnapshotCache, SnapshotPointer, ThrottleOptions,
+};
+
+////////////////////////////////////////////// Garbage /////////////////////////////////////////////
+
+#[derive(Debug, Clone, Default)]
+pub struct Garbage {
+    pub dropped_setsum: Setsum,
+    pub actions: Vec<GarbageAction>,
+    pub cutoff: LogPosition,
+}
+
+impl Garbage {
+    #[allow(clippy::result_large_err)]
+    pub async fn new<C: SnapshotCache>(
+        storage: &Storage,
+        prefix: String,
+        manifest: &Manifest,
+        throttle: &ThrottleOptions,
+        snapshots: &C,
+        first_to_keep: LogPosition,
+    ) -> Result<Self, Error> {
+        let dropped_snapshots = manifest
+            .snapshots
+            .iter()
+            .filter(|snap| snap.limit <= first_to_keep)
+            .collect::<Vec<_>>();
+        let replaced_snapshots = manifest
+            .snapshots
+            .iter()
+            .filter(|snap| (snap.start..snap.limit).contains(&first_to_keep))
+            .collect::<Vec<_>>();
+        let dropped_fragments = manifest
+            .fragments
+            .iter()
+            .filter(|frag| frag.limit <= first_to_keep)
+            .collect::<Vec<_>>();
+        let mut actions = vec![];
+        let mut drop_acc = Setsum::default();
+        for frag in dropped_fragments {
```

Comment:
qq: can these operations be applied in any order to the setsum i.e. is the setsum associative?